pipeline/store/accessor.go

```diff
@@ -0,0 +1,177 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/streamingfast/dstore"
+	"github.com/streamingfast/substreams/block"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+	"github.com/streamingfast/substreams/state"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var ErrStoreNotFound = errors.New("store not found")
+
+type Accessor struct {
+	baseStore           dstore.Store
+	storeMap            map[string]*state.Store
+	SaveInterval        uint64
+	maxBoundaryBlockNum uint64
+	nextSaveBoundary    uint64
+	logger              *zap.Logger
+}
+
+func NewAccessor(saveInterval, minBoundaryBlock, maxBoundaryBlockNum uint64, baseStore dstore.Store, logger *zap.Logger) *Accessor {
+	s := &Accessor{
+		storeMap:            map[string]*state.Store{},
+		SaveInterval:        saveInterval,
+		maxBoundaryBlockNum: maxBoundaryBlockNum,
+		baseStore:           baseStore,
+		logger:              logger,
+	}
+
+	s.nextSaveBoundary = s.computeBoundaryBlock(minBoundaryBlock)
+	return s
+}
+
+func (s *Accessor) AddStore(hash string, storeModule *pbsubstreams.Module) (err error) {
+	store, err := state.NewStore(
+		storeModule.Name,
+		s.SaveInterval,
+		storeModule.InitialBlock,
+		// gets thme for store
+		hash,
+		storeModule.GetKindStore().UpdatePolicy,
+		storeModule.GetKindStore().ValueType,
+		s.baseStore,
+		s.logger,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to load store: %w", err)
+	}
+	s.storeMap[storeModule.Name] = store
+	return nil
+}
+
+func (s *Accessor) Get(storeName string) (*state.Store, error) {
+	if store, found := s.storeMap[storeName]; found {
+		return store, nil
+	}
+	return nil, ErrStoreNotFound
+}
+
+func (s *Accessor) Update(storeName string, newStore *state.Store) bool {
+	if _, found := s.storeMap[storeName]; found {
+		s.storeMap[storeName] = newStore
+		return true
+	}
+	return false
+}
+
+func (s *Accessor) Hydrate(ctx context.Context, blockNum uint64) error {
+	for _, store := range s.storeMap {
+		if store.StoreInitialBlock() == blockNum {
+			continue
+		}
+
+		err := store.Fetch(ctx, blockNum)
+		if err != nil {
+			return fmt.Errorf("failed to hydrate store %q: %w", store.Name, err)
+		}
+	}
+	return nil
+}
+
+func (s *Accessor) ResetStores() {
+	for _, store := range s.storeMap {
+		store.Reset()
+	}
+}
+
+// will attempt tp flush the stores  (or stores in filter if present) and report back the saved ranges
+func (r *Accessor) Flush(ctx context.Context, blockNum uint64, stopBlockNum uint64, storeFilter map[string]bool) (block.Ranges, error) {
+	var out block.Ranges
+	if err := r.flushLoop(blockNum, stopBlockNum, func() error {
+		blockRanges, err := r.flushStores(ctx, r.nextSaveBoundary, storeFilter)
+		if err != nil {
+			return fmt.Errorf("failed to flush stores: %w", err)
+		}
+		out = append(out, blockRanges...)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
```

Comment:
All this work! `Hydrate`, `Flush` .. which is business logic of the user.

I think we can give another run of analysis on that famous Accessor :)